pandera/engines/engine.py

```diff
@@ -0,0 +1,157 @@
+        ):
+            raise ValueError(
+                f"{cls._base_datatype.__name__} subclasses cannot be registered"
+                f" with {cls.__name__}."
+            )
+
+    def _register_from_parametrized_dtype(
+        cls, target_dtype: Union[DataType, Type[DataType]], method: classmethod
+    ) -> None:
+        func = method.__func__
+        annotations = get_type_hints(func).values()
+        dtype = next(iter(annotations))  # get 1st annotation
+        # parse typing.Union
+        dtypes = typing_inspect.get_args(dtype) or [dtype]
+
+        def _method(*args, **kwargs):
+            return func(target_dtype, *args, **kwargs)
+
+        for source_dtype in dtypes:
+            cls._check_source_dtype(source_dtype)
+            cls._registry[cls].dispatch.register(source_dtype, _method)
+
+    def _register_equivalents(
+        cls,
+        target_dtype: Union[DataType, Type[DataType]],
+        *source_dtypes: Any,
+    ) -> None:
+        value = target_dtype()
+        for source_dtype in source_dtypes:
+            cls._check_source_dtype(source_dtype)
+            cls._registry[cls].equivalents[source_dtype] = value
+
+    def register_dtype(
+        cls,
+        dtype: Type[DataType] = None,
+        *,
+        equivalents: List[DataType] = None,
+    ):
+        """Register a DataType
+
+        :param dtype: The DataType to register.
+        :param equivalents: Equivalent scalar dtype class or
+            non-parametrized dtype instance.
+
+        .. note::
+            Register the classmethod ``from_parametrized_dtype`` if present.
+        """
+
+        def _wrapper(dtype: Union[DataType, Type[DataType]]):
+            if not inspect.isclass(dtype):
+                raise ValueError(
+                    f"{cls.__name__}.register_dtype can only decorate a class, "
+                    + f"got {dtype}"
+                )
+
+            if equivalents:
+                cls._register_equivalents(dtype, *equivalents)
+
+            from_parametrized_dtype = dtype.__dict__.get(
+                "from_parametrized_dtype"
+            )
+            if from_parametrized_dtype:
+                if not isinstance(from_parametrized_dtype, classmethod):
+                    raise ValueError(
+                        f"{dtype.__name__}.from_parametrized_dtype "
+                        + "must be a classmethod."
+                    )
+                cls._register_from_parametrized_dtype(
+                    dtype,
+                    from_parametrized_dtype,
+                )
+            elif not equivalents:
+                warnings.warn(
+                    f"register_dtype({dtype}) on a class without a "
+                    + "'from_parametrized_dtype' classmethod has no effect."
+                )
+
+            return dtype
+
+        if dtype:
+            return _wrapper(dtype)
+
+        return _wrapper
+
+    def dtype(cls, obj: Any) -> DataType:
```

Comment:
can we rename `obj` to `data_type` or `native_data_type`?  Was confused here a bit since `obj` is used elsewhere as the object to coerce

Reply:
Ok.